Alex and Happy Part 1/article.md

```diff
@@ -0,0 +1,842 @@
+
+This two-part tutorial will look into two tools often used together by Haskellers to parse programs: Alex and Happy.
+We'll use them to parse a small programming language from scratch.
+
+Both Alex and Happy are industrial-strength tools powerful enough to parse even Haskell itself.
+At the bottom of this tutorial, you will find links to GHC's lexer and parser if you are curious to see how they look.
+
+This tutorial was written using GHC version 9.0.2, Stack resolver LTS-19.8, Alex version 3.2.7.1, and Happy version 1.20.0.
+
+## Our grammar: MiniML
+
+<!--
+* [X] describe what we support
+* [X] example snippet
+-->
+
+For this tutorial, we will introduce a small grammar called MiniML (read as "minimal") that's based on [ML](https://en.wikipedia.org/wiki/ML_(programming_language)) .
+As the name suggests, it has the minimal syntax needed to get you started with Alex and Happy. Yet we hope to introduce enough concepts for you to be able to create a useful grammar for your programming language and use these tools effectively.
+
+Some of the features that this tutorial series will support are:
+* Variables, constants, function declarations, function applications, local bindings, and operators.
+* Some simple types, as well as arrow types (functions).
+* Conditional expressions: if-then-else and if-then.
+* Block comments.
+
+In addition, exercises will guide you through extending the language with line comments and rational numbers. Later, in the second part, exercises will introduce patterns, list access, and pattern matching.
+
+The snippet below demonstrates a program written in MiniML:
+
+```sml
+let the_answer : int =
+  let a = 20 in
+  let b = 1 in
+  let c = 2 in
+  a * c + b * c
+
+let main (unit : ()) : () =
+  print ("The answer is: " + the_answer)
+```
+
+At the end of the parsing process, we will build an _abstract syntax tree_ (AST), which is a data structure representing a program.
+In contrast to unstructured data (like a string), an AST is much easier to manipulate.
+You can use it to write an interpreter, transpile into other language, etc.
+
+Here's how the (simplified and prettified) AST of the snippet above will look at at the end of the second part:
+
+```haskell
+[ Dec _ (Name _ "the_answer") [] (Just (TVar _ (Name _ "int")))
+  (ELetIn _ (Dec _ (Name _ "a") [] Nothing (EInt _ 20))
+  (ELetIn _ (Dec _ (Name _ "b") [] Nothing (EInt _ 1))
+  (ELetIn _ (Dec _ (Name _ "c") [] Nothing (EInt _ 2))
+  (EBinOp _ (EBinOp _ (EVar _ (Name _ "a")) (Times _) (EVar _ (Name _ "c"))) (Plus _) (EBinOp _ (EVar _ (Name _ "b")) (Times _) (EVar _ (Name _ "c")))))))
+, Dec _ (Name _ "main") [Argument _ (Name _ "unit") (Just (TUnit _))] (Just (TUnit _))
+  (EApp _ (EVar _ (Name _ "print")) (EPar _ (EBinOp _ (EString _ "\"The answer is: \"") (Plus _) (EVar _ (Name _ "the_answer")))))
+]
+```
+
+n.b.: The `_` will contain the range of the parsed tree, we've omited it here for brevitity.
```

Comment:
I have a feeling like it should be the ranges of the parsed tokens, since every token has a range, am I wrong?